bpf/lib/lb.h

```diff
@@ -1748,8 +1752,341 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	__be32	saddr;
+	__be32	daddr;
+	__u8	tos;
+	__wsum csum;
+	int sample_len;
+	const int inner_offset = sizeof(struct ethhdr) + sizeof(struct iphdr) +
+		sizeof(struct icmphdr);
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	/* copy the incoming src and dest IPs and mac addresses to the stack.
+	 * the pointers will not be valid after adding headroom.
+	 */
+
+	if (eth_load_saddr(ctx, smac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	if (eth_load_daddr(ctx, dmac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	saddr = ip4->saddr;
+	daddr = ip4->daddr;
+	tos = ip4->tos;
+
+	/* Resize to ethernet header + 64 bytes or less */
+	sample_len = ctx_full_len(ctx);
+	if (sample_len > ICMP_PACKET_MAX_SAMPLE_SIZE)
+		sample_len = ICMP_PACKET_MAX_SAMPLE_SIZE;
+	ctx_adjust_troom(ctx, sample_len + sizeof(struct ethhdr) - ctx_full_len(ctx));
+
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	/* Unrolled loop to calculate the checksum of the ICMP sample
+	 * Done manually because the compiler refuses with #pragma unroll
+	 */
+	csum = 0;
+
+	#define body(i) if ((i) > sample_len) \
+			goto csumend; \
+		if (data + sizeof(struct ethhdr) + (i) + sizeof(__u16) > data_end) \
+			goto csumend; \
+		csum += *(__u16 *)(data + sizeof(struct ethhdr) + (i));
+
+	#define body5(i) body(i)\
+		body(i + 2) \
+		body(i + 4) \
+		body(i + 6) \
+		body(i + 8)
+
+	body5(0)
+	body5(10)
+	body5(20)
+	body5(30)
+	body5(40)
+	body5(50)
+	body(60)
+	body(62)
+
+csumend :
+
+	/* We need to insert a IPv4 and ICMP header before the original packet.
+	 * Make that room.
+	 */
+
+#if __ctx_is == __ctx_xdp
+	xdp_adjust_head(ctx, 0 - (int)(sizeof(struct iphdr) + sizeof(struct icmphdr)));
+#else
+	skb_adjust_room(ctx, sizeof(struct iphdr) + sizeof(struct icmphdr), BPF_ADJ_ROOM_MAC, 0);
+#endif
+
+	/* changing size invalidates pointers, so we need to re-fetch them. */
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+
+	/* Bound check all 3 headers at once. */
+	if (data + inner_offset > data_end)
+		return DROP_INVALID;
+
+	/* Write reversed eth header, ready for egress */
+	ethhdr = data;
+	memcpy(ethhdr->h_dest, smac.addr, sizeof(smac.addr));
+	memcpy(ethhdr->h_source, dmac.addr, sizeof(dmac.addr));
+	ethhdr->h_proto = bpf_htons(ETH_P_IP);
+
+	/* Write reversed ip header, ready for egress */
+	ip4 = data + sizeof(struct ethhdr);
+	ip4->version = 4;
+	ip4->ihl = sizeof(struct iphdr) >> 2;
+	ip4->tos = tos;
+	ip4->tot_len = bpf_htons(sizeof(struct iphdr) + sizeof(struct icmphdr) +
+		       (__u16)sample_len);
+	ip4->id = (__be16)get_prandom_u32();
+	ip4->frag_off = 0;
+	ip4->ttl = 64;
+	ip4->protocol = IPPROTO_ICMP;
+	ip4->check = 0;
+	ip4->daddr = saddr;
+	ip4->saddr = daddr;
+	ip4->check = csum_fold(csum_diff(ip4, 0, ip4, sizeof(struct iphdr), 0));
+
+	/* Write reversed icmp header */
+	icmphdr = data + sizeof(struct ethhdr) + sizeof(struct iphdr);
+	icmphdr->type = ICMP_DEST_UNREACH;
+	icmphdr->code = ICMP_HOST_UNREACH;
```

Comment:
My quick thought here is that a Service K8s resource is defined as being associated with a port, so it would be important to NOT use host unreachable, as the IP is likely to be shared among multiple services and/or reused within some cache interval.

Reply:
That is a good point, I don't know if this is cached somehow. Agreed, lets stick to port unreachable for now since we know that works for others.